### ISSUE TITLE
db: Add Landuse category to database

### ISSUE THREAD

**Issue by dominijk** [colouring-london/colouring-london#358]
**Fields for launch**
~~- Single or multiple use > Integer if more than 1 opens multi entry forms for number given~~
- Default is to add one land use then as with age category `url` link enable more to be added giving user prompt and guidance of how to do this  
- Class which auto populates remaining group and order based on class field . This links to #363 as class can contain descriptor of attachment ie. Detached house @tomalrussell is there anything react can do to take that value and flag to user that it's prefilled the field in the categories? 
- Add source
- Has verification button and review button [longer term feature launch] 

**Steps for adding new fields to database;**

- [x] Identify bulk data sources (Function Sties and London Development Database)
- [x] Required - Define field values as static category page
- [ ] Required - Identify need for source and verification for fields, log to spreadsheet
- [x] Required - Define input method for field (Drop down / Autocomplete with autofill?) 
- [x] Required - Code category down file
- [x] Required - Code category up file
- [x] Required - Test category down file
- [x] Required - Test category up file
- [x] Required - Add category down file
- [x] Required - Add category up file
- [ ] On completion set up Github issue for activating fields on front end #513

**Comment by dominijk** [comment]
Working on autofill for linked fields. Ie. Input in field `a` updates field `b`  
- Can/should this happen in the client. If it doesn't then has to wait for data save to run on server 
- How would it work for mixed use (if number of uses is >1 then mixed, generate forms for number uses specified )

**Comment by tomalrussell** [comment]
Shall we introduce a table or three for the order and group here?
- descriptions can be as long as we like
- table design could include short id and parent/child relationship


| landuse_id | description | level | parent |
| ---- | ---- | ---- | ---- | 
| U010 | Agriculture and Fisheries | order | null |
| U011 | Agriculture | group | U010 |
| U012 | Fisheries | group | U010 |

Is there a third level too? Are these the best docs? https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/11493/144275.pdf

**Comment by tomalrussell** [comment]
> Can/should this happen in the client. If it doesn't then has to wait for data save to run on server, so on client lock the field and say this will be completed automagically.

Yes, should all be client side, keeping the server as simple as possible.

> How would it work for mixed use (if number of uses is >1 then mixed, generate forms for number uses specified ) would this be a `JavaScript switch` [statement](https://www.w3schools.com/js/js_switch.asp)?

I think this would be a multi-entry form, something like the approach used for 'Text and Image Links' under age, so if there's a single use, it's a list of one use, if there are multiple uses, it's a list of more than one.

For reference, the MDN docs (e.g. [`switch`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch)) are typically high quality - would recommend over w3schools.

> ~For non mixed this could trim after hyphen ie. `class` Studio - Offices  is trimmed to `order` Offices , can then use switch to look up the 41 `orders` to the  13 `Groups`~ Enumerated types limited to 63 char, so input in browser ok but will bounce at db unless you strip after hyphen

It may be worth upgrading from enum to table here - see note above. Or the ENUMs could use only the short id (`U010`), and the dropdown/autocomplete with description could be entirely in the client.

**Comment by dominijk** [comment]
@tomalrussell Thanks, happy to go with your suggestion on tables there are three levels to the hierachy 
- landuse_class
- landuse_group
- landuse_order

Which should be in the above commit for reference, those are extracted from that document as noted in the commit. On my local copy I've also started a table called, this is for my internal reference but perhaps we should something similar to server as part of the field source development?

Re: Client versus server side if we wanted to have triggers on the db for anything is that an option? 

|source_name|source_location|db_table_name|
|-----------|---------------|-----------------|
|gov_nlud_classifications|https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/11493/144275.pdf|nlud_classification|
|historic_england_nhle|email_spreadsheet||

**Comment by tomalrussell** [comment]
Great
- let's put them in a single table with a `level` column which is one of 'order', 'group' or 'class'
- let's tackle field sources as part of that issue - a table might be useful, but we're more likely to want free text to cover potential variety
- it would be useful to surface these kinds of sources, and to document how we're using them - the Wix site and/or the GitHub wiki would be the places for less/more technical reference.
- on triggers, I would strongly recommend against doing anything implicit to the data behind the API: like OpenStreetMap, we should assume that local knowledge and community contributions take top priority 
  - things that we can do client-side can support collecting data across categories and support user contributions by making suggestions, but should always be transparent and explicitly approved - never automatically updating or overwriting.
  - if we - or others - want to programmatically set e.g. building material for all buildings in an age band, that's a good use case for a script that works against a bulk download CSV and posts updates to the site. Those updates get logged in the edit history; the script can run when appropriate and be improved independently from anything going on in the simple CL server.

**Comment by dominijk** [comment]
Thanks that’s a nice approach, they’re already on the wix but I can document on the wiki as well.

**Comment by dominijk** [comment]
@tomalrussell On this thread does table make it easier to add new fields, which links to #284 where people can suggest another value for a field which can then get held until that value is moderated as being a valid addition to the list (ie approved source for historic data). For launch it could simply be link to suggest an extra value which can be manually run.

**Comment by tomalrussell** [comment]
Notes from discussion on handling validation for this on the server-side.

In the building-update handler, before we hit the database to update it, run checks:
- are we updating landuse_class? - if not, continue
- are the classes all valid? - if not, error
- are the classes all from the same group? - if yes, set group, else set mixed
- continue onwards to database-saving code

**Comment by mz8i** [comment]
@tomalrussell would it make sense to implement this as a database trigger? On update of `current_landuse_class` or `original_landuse_class`, calculate and update `current_landuse_order` or `original_landuse_order`, respectively. I'm unsure about putting this kind of conditional domain-related logic in the API right now.

**Comment by tomalrussell** [comment]
I'd prefer not to put any logic of this kind in the database, and I was previously a little unsure about putting too much in the API too.

Can we handle it all on the front end?

Another idea, which has different disadvantages: perhaps we could also write an external validation bot that could run nightly and process the edit history to update any mismatched class/order pairs?

**Comment by mz8i** [comment]
I have since realised that putting the logic in the database would lead to potential inconsistencies in the edit history, since that is handled solely on the API level. Just now working on adding the processing to the API.

My personal feeling reflects what you said in #488, that we need to be consistent about validation. Maybe some constraints can be only db+front, and some only api+front, but I don't feel frontend-only constraints are an option as the client should usually be treated as an unknown, someone can be just hitting the API. Let's maybe discuss for the future in #524 that I created recently
